File: src/main/java/org/parsimonygroup/FunctionBootstrap.java
package org.parsimonygroup;

import cascading.flow.FlowProcess;
import cascading.operation.BaseOperation;
import cascading.operation.Function;
import cascading.operation.FunctionCall;
import cascading.tuple.Fields;
import cascading.tuple.Tuple;
import cascading.tuple.TupleEntry;
import cascading.tuple.TupleEntryCollector;
import clojure.lang.IFn;

import java.util.Collection;

public class FunctionBootstrap extends BaseOperation implements Function {
  private IFn reader;
  private IFn function;
  private ClojureCascadingHelper clojureHelper;
  private IFn writer;
  private IFn cljCallback;

  public FunctionBootstrap(Fields inFields, Fields outFields, IFn reader, IFn writer, IFn function, IFn cljCallback, String fnNsName) {
    super( outFields);
    this.reader = reader;
    this.function = function;
    this.writer = writer;
    this.cljCallback = cljCallback;
    this.clojureHelper = new ClojureCascadingHelper(fnNsName);
  }

  public void operate(FlowProcess flowProcess, FunctionCall functionCall) {
    try {
      clojureHelper.bootClojure();
    } catch (Exception e) {
      e.printStackTrace();
    }
    TupleEntryCollector outputCollector = functionCall.getOutputCollector();
    TupleEntry arguments = functionCall.getArguments();
    processData(arguments, outputCollector);
  }

  private void processData(TupleEntry arguments, TupleEntryCollector collector) {
    try {
      Collection<Tuple> resultTuples = clojureHelper.callClojure(arguments, function, cljCallback, reader, writer);
      for(Tuple tuple : resultTuples) {
         collector.add(tuple);
      }
    } catch (Exception e) {
      throw new RuntimeException("error prcessing Data", e);
    }
  }
}

File: src/main/clojure/org/parsimonygroup/function_bootstrap.clj
(ns org.parsimonygroup.function-bootstrap
  (:import [cascading.flow FlowProcess]
           [cascading.operation BaseOperation Function FunctionCall]
           [cascading.tuple Fields Tuple TupleEntryCollector TupleEntry])
  (:use org.parsimonygroup.utils))


(gen-class
 :name org.parsimonygroup.FunctionBootstrap
 :extends cascading.operation.BaseOperation
 :implements [cascading.operation.Function]
 :constructors {[cascading.tuple.Fields cascading.tuple.Fields
                 clojure.lang.IFn clojure.lang.IFn
                 clojure.lang.IFn clojure.lang.IFn String]
                [cascading.tuple.Fields]}
 :init init
 :state state)


;; Oddly Cascading gags when trying to serialize
;; Clojure keywords so we can't use them here:
(defn -init [in-fields out-fields reader writer function
             clj-callback fn-ns-name]
  [[out-fields] {"reader" reader "writer" writer "function" function
                 "clj-callback" clj-callback
                 "fn-ns-name" fn-ns-name}])

(defn -prepare [this _ _]
  (boot-clojure (.state this)))


(defn process-data [this arguments collector]
   (let [state (.state this)
         result-tuples
         ((state "clj-callback" )
          (state "reader" ) (state "writer" )
          (state "function" )  (iterator-seq (.. arguments getTuple iterator)))
         result-tuples (map #(Tuple. (into-array Comparable %)) result-tuples)]
     (doseq [tuple result-tuples]
       (.add collector tuple))))

(defn -operate [this flow-process function-call]
  (let [output-collector (.getOutputCollector function-call)
        arguments (.getArguments function-call)]
    (process-data this arguments output-collector)))


